windows/devices/uwb/UwbConnector.cxx

```diff
@@ -915,8 +939,90 @@ UwbConnector::CallbacksPresent()
 }
 
 void
-UwbConnector::DeregisterEventCallback(::uwb::RegisteredCallbackToken* token)
+UwbConnector::DeregisterEventCallback(std::weak_ptr<::uwb::RegisteredCallbackToken> token)
 {
-    // TODO implement
+    auto tok = token.lock();
+    if (not tok) {
+        return;
+    }
     std::lock_guard eventCallbacksLockExclusive{ m_eventCallbacksGate };
+
+    DeregisterEventCallback(*tok);
 }
+
+void
+UwbConnector::DeregisterEventCallback(::uwb::RegisteredCallbackToken token)
+{
+    auto callbackId = token.callbackId;
+    auto isDeviceEventCallback = token.isDeviceEventCallback;
+
+    // first remove the token from the list of tokens
+    for (auto it = std::cbegin(m_tokens); it != std::cend(m_tokens); it++) {
+        if ((*it)->callbackId == callbackId) {
+            m_tokens.erase(it);
+            break;
+        }
+    }
+
+    // deal with the id map and list of callbacks
+    if (isDeviceEventCallback) {
+        // first remove the id from the id map
+        auto result = m_deviceEventCallbacksIdMap.find(callbackId);
+        if (result == std::cend(m_deviceEventCallbacksIdMap)) {
+            return;
+        }
+        m_deviceEventCallbacksIdMap.erase(result);
+
+        // now remove the id from the list of callbacks
+        for (auto it = std::cbegin(m_deviceEventCallbacks); it != std::cend(m_deviceEventCallbacks); it++) {
+            if (*it == callbackId) {
+                m_deviceEventCallbacks.erase(it);
+                return;
+            }
+        }
+    } else {
+        // first remove the id from the id map
+        auto result = m_sessionEventCallbacksIdMap.find(callbackId);
+        if (result == std::cend(m_sessionEventCallbacksIdMap)) {
+            return;
+        }
+        m_sessionEventCallbacksIdMap.erase(result);
+
+        // now remove the id from the list of callbacks
+        for (auto sessionIt = std::cbegin(m_sessionEventCallbacks); sessionIt != std::cend(m_sessionEventCallbacks); sessionIt++) {
+            auto [sessionId, sessionCallbackIds] = *sessionIt;
+            if (sessionCallbackIds.empty()) {
+                PLOG_ERROR << "empty sessionEventCallbacks for session with id=" << sessionId;
+                continue;
+            }
+            for (auto it = std::cbegin(sessionCallbackIds); it != std::cend(sessionCallbackIds); it++) {
+                if (*it == callbackId) {
+                    sessionCallbackIds.erase(it);
+                    // if this is the last sessionEventCallback for this sessionId, make sure to erase the entire vector from m_sessionEventCallbacks
+                    if (sessionCallbackIds.empty()) {
+                        m_sessionEventCallbacks.erase(sessionIt);
+                    }
+                    return;
+                }
+            }
+        }
+    }
+}
+
+std::shared_ptr<::uwb::RegisteredCallbackToken>
+UwbConnector::InsertDeviceEventCallback(std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks> callback)
+{
+    m_tokenUniqueState++;

```

Comment:
This isn't clear. What is the unique state and why is it incremented as an integer?

Reply:
m_tokenUniqueState is just a (admittedly, hacky) way to maintain unique ids for each callback that gets registered. Each time a callback is registered, it gets assigned an id of m_tokenUniqueState and then m_tokenUniqueState is incremented.

Reply:
I see. In that case, I think the name is misleading since the variable is meant to track a unique ID number but the name implies that it's tracking some sort of state.